### PR TITLE
[Do not Mege] Updated Set-AzureRmVMChefExtension cmd

### PR DESCRIPTION
**Pull request by NimishaS**
1. Using new classes like `CreateOrUpdateWithHttpMessagesAsync`
2. Passing settings file as Hashtable
3. Implements `Get-AzureRmVMChefExtension`
4. Implements `Remove-AzureRmVMChefExtension`

Testing is in progress

### CODE REVIEW

**Comment by azurecla** [comment]
Hi __@NimishaS__, I'm your friendly neighborhood Azure Pull Request Bot (You can call me AZPRBOT). Thanks for your contribution!
    <span>You've already signed the contribution license agreement. Thanks!</span>
        <span>We will now validate the agreement and then real humans will evaluate your PR.</span>

TTYL, AZPRBOT;

**Comment by hovsepm** [comment]
@NimishaS Please remove [Do not Merge] from the title when your tests will be in the PR.

**Comment by NimishaS** [comment]
@hovsepm , can you point me to a doc which explains how to run a specific test scenario. Eg. suppose I want to run only https://github.com/Azure/azure-powershell/blob/dev/src/ResourceManager/Compute/Commands.Compute.Test/ScenarioTests/DiagnosticsExtensionTests.ps1
This wiki https://github.com/Azure/azure-powershell/wiki/Microsoft-Azure-PowerShell-Developer-Guide#run-unit-tests explains how to run all unit test-cases.

**Comment by hovsepm** [comment]
@NimishaS to run any specific test you will need to first find a corresponding C# Unit test that executes that script and run it with Xunit framework (usually it is much easier to do in Visual Studio).  
Here is the unit test that invokes Ps script you are asking about - https://github.com/Azure/azure-powershell/blob/3d52731a95aaf1cbb6e04ce255b7d4958e6353d1/src/ResourceManager/Compute/Commands.Compute.Test/ScenarioTests/DiagnosticsExtensionTests.cs#L24

**Comment by NimishaS** [comment]
@hovsepm , we are done adding the specs. We are able to run the tests in `Playback` mode but not in `Recording` mode. We tried setting `AZURE_TEST_MODE` to `RECORD` in env variables. But the tests still ran in `Playback` mode from `Visual Studio`.
Tried to use `Microsoft Test Manager` but we have to add `Team Foundation Server` here. We are not sure what value is supposed to be given here.
These steps are not helping us for recording https://github.com/Azure/azure-powershell/wiki/Microsoft-Azure-PowerShell-Developer-Guide#recordingrunning-tests
Please suggest.

**Comment by NimishaS** [comment]
Also Travis is failing for this PR but we are unable to see the details.

**Comment by NimishaS** [comment]
@hovsepm , we are finally able to run the tests in `recording` mode. But we are getting this error for each spec:
```
System.Exception : Logged in account had no associated subscriptions. We are in Prod environment
and the tenant is common. Please check if the subscription is in the correct tenant and environment.
You can set the envt. variable - TEST_CSM_ORGID_AUTHENTICATION=SubscriptionId=<subscription-id>;Environment=<Env-name>;Tenant=<tenant-id>
```
We have set the environment variables following https://github.com/Azure/azure-powershell/wiki/Microsoft-Azure-PowerShell-Developer-Guide#ad-scenario-tests.
There is a possibility that the values provided here are not correct. We are not able to understand if the values provided for `UserId` in the four variables are same or different.

**Comment by NimishaS** [comment]
@hovsepm , any suggestions here?

**Comment by NimishaS** [comment]
@hovsepm we are waiting for your response to record test responses. Please refer my comment https://github.com/Azure/azure-powershell/pull/1676#issuecomment-175615995. That's the problem we are currently stuck on.

Rest everything is complete for this PR.

**Comment by hovsepm** [comment]
@NimishaS the issue is that you did not include recorded mock recordings to the PR.
```
Microsoft.Azure.Commands.Compute.Test.ScenarioTests.ChefExtensionTests.TestChefExtensionBasic
System.ArgumentException : Unable to find recorded mock directory 'SessionRecords\Microsoft.Azure.Commands.Compute.Test.ScenarioTests.ChefExtensionTests'.
Parameter name: callerIdentity
   at Microsoft.Azure.Test.HttpRecorder.HttpMockServer.Initialize(String callerIdentity, String testIdentity, HttpRecorderMode mode)
   at Microsoft.Rest.ClientRuntime.Azure.TestFramework.MockContext.Start(String className, String methodName)
   at Microsoft.Azure.Commands.Compute.Test.ScenarioTests.ComputeTestController.RunPsTestWorkflow(Func`1 scriptBuilder, Action`1 initialize, Action cleanup, String callingClassType, String mockName) in d:\workspace\powershell\src\ResourceManager\Compute\Commands.Compute.Test\Common\ComputeTestController.cs:line 105
   at Microsoft.Azure.Commands.Compute.Test.ScenarioTests.ComputeTestController.RunPsTest(String[] scripts) in d:\workspace\powershell\src\ResourceManager\Compute\Commands.Compute.Test\Common\ComputeTestController.cs:line 84
   at Microsoft.Azure.Commands.Compute.Test.ScenarioTests.ChefExtensionTests.TestChefExtensionBasic() in d:\workspace\powershell\src\ResourceManager\Compute\Commands.Compute.Test\ScenarioTests\ChefExtensionTests.cs:line 12
Traits:
```

**Comment by hovsepm** [comment]
@NimishaS please see SessionRecords folder for an example. You will need to add your session record json file there in the correct folder, and add that file to Commands.Compute.Test.csproj. Make sure CopyToOutputDirectory setting will me set to "Always" in the csproj.

File: src/ResourceManager/Compute/Commands.Compute.Test/ScenarioTests/ChefExtensionTests.ps1 [hunk 1]
```diff
@@ -0,0 +1,45 @@
...
+    $loc = Get-ComputeVMLocation
+
+    try
+    {
+        # Create Virtual Machine
+        $vm = Create-VirtualMachine -rgname $rgname -loc $loc
+        $vmname = $vm.Name
+		$version = "1210.12"
+		$client_rb = ".\Templates\client.rb";
+		$validationPemFile = ".\Templates\tstorgnztn-validator.pem";
```
Comment (hovsepm) [inline src/ResourceManager/Compute/Commands.Compute.Test/ScenarioTests/ChefExtensionTests.ps1]: please align the white-spaces in this file.

**Comment by NimishaS** [comment]
@hovsepm , issue is that we aren't able to record specs. When we run the tests in `recording` mode, we get this error:
```
System.Exception : Logged in account had no associated subscriptions. We are in Prod environment 
and the tenant is common. Please check if the subscription is in the correct tenant and environment. 
You can set the envt. variable - 
TEST_CSM_ORGID_AUTHENTICATION=SubscriptionId=<subscription-id>;Environment=<Env-name>;Tenant=<tenant-id>
```
We get this error for all specs. We have tried setting the environment variables mentioned here https://github.com/Azure/azure-powershell/wiki/Microsoft-Azure-PowerShell-Developer-Guide#ad-scenario-tests

**Comment by markcowl** [comment]
@NimishaS Is this a tenant-only RP?  Updating the test framework should provide support for tenant-specific RPs.  Note that, if you use clients in your tests that require a subscription id (like resource manager), then this issue will still hold - you need to make sure that clients and cmdlets used in your tests do not require subscriptionId.

**Comment by markcowl** [comment]
@NimishaS You sho9uld set TEST_CSM_ORGID_AUTHENTICATION, as specified in the error message - if you specify SubscriptionId and Environment, you will be prompted for credentials during the test.

**Comment by NimishaS** [comment]
@markcowl, TEST_CSM_ORGID_AUTHENTICATION is set. Still we are getting the same error.

**Comment by markcowl** [comment]
@NimishaS What help do you need to get these changes in?  We will have to close this pull request soon, as it is quite out of date.

**Comment by markcowl** [comment]
@NimishaS closing this PR, as there has been no activity in 2 months.  Open a new PR if you weant us to consider changes.

**Pull request closed without merging**